File: .UNSAFE-ALL-BINDINGS-DELETED-20260125-095646/bindings/scala/src/main/scala/proven/SafeAngle.scala
// SPDX-License-Identifier: PMPL-1.0

package proven

/**
 * Angle representation with safe conversions.
 */
case class Angle private (radians: Double):

  /**
   * Get value in degrees.
   */
  def degrees: Double = radians * 180.0 / Math.PI

  /**
   * Get value in gradians (grads).
   */
  def gradians: Double = radians * 200.0 / Math.PI

  /**
   * Get value in turns (full rotations).
   */
  def turns: Double = radians / (2 * Math.PI)

  /**
   * Normalize to [0, 2*PI).
   */
  def normalize: Angle =
    val normalized = radians % (2 * Math.PI)
    Angle(if normalized < 0 then normalized + 2 * Math.PI else normalized)

  /**
   * Normalize to [-PI, PI).
   */
  def normalizeSigned: Angle =
    val normalized = normalize.radians
    Angle(if normalized >= Math.PI then normalized - 2 * Math.PI else normalized)

  /**
   * Add two angles.
   */
  def +(other: Angle): Angle = Angle(radians + other.radians)

  /**
   * Subtract two angles.
   */
  def -(other: Angle): Angle = Angle(radians - other.radians)

  /**
   * Multiply angle by scalar.
   */
  def *(scalar: Double): Angle = Angle(radians * scalar)

  /**
   * Divide angle by scalar.
   */
  def /(scalar: Double): Option[Angle] =
    if scalar.abs < 1e-10 then None
    else Some(Angle(radians / scalar))

  /**
   * Negate the angle.
   */
  def unary_- : Angle = Angle(-radians)

  /**
   * Sine of the angle.
   */
  def sin: Double = Math.sin(radians)

  /**
   * Cosine of the angle.
   */
  def cos: Double = Math.cos(radians)

  /**
   * Tangent of the angle.
   */
  def tan: Option[Double] =
    val normalized = normalize.radians
    // tan is undefined at PI/2 and 3*PI/2
    if (normalized - Math.PI / 2).abs < 1e-10 || (normalized - 3 * Math.PI / 2).abs < 1e-10 then
      None
    else
      Some(Math.tan(radians))

  /**
   * Check if angle is acute (< 90 degrees).
   */
  def isAcute: Boolean =
    val norm = normalize.radians
    norm > 0 && norm < Math.PI / 2

  /**
   * Check if angle is right (= 90 degrees).
   */
  def isRight: Boolean =
    (normalize.radians - Math.PI / 2).abs < 1e-10

  /**
   * Check if angle is obtuse (90 < angle < 180 degrees).
   */
  def isObtuse: Boolean =
    val norm = normalize.radians
    norm > Math.PI / 2 && norm < Math.PI

  /**
   * Check if angle is straight (= 180 degrees).
   */
  def isStraight: Boolean =
    (normalize.radians - Math.PI).abs < 1e-10

  /**
   * Check if angle is reflex (> 180 degrees).
   */
  def isReflex: Boolean =
    val norm = normalize.radians
    norm > Math.PI && norm < 2 * Math.PI

  override def toString: String = f"${degrees}%.2f deg"

object Angle:
  val Zero: Angle = Angle(0)
  val RightAngle: Angle = Angle(Math.PI / 2)
  val StraightAngle: Angle = Angle(Math.PI)
  val FullRotation: Angle = Angle(2 * Math.PI)

  /**
   * Create angle from radians.
   */
  def fromRadians(rad: Double): Angle = new Angle(rad)

  /**
   * Create angle from degrees.
   */
  def fromDegrees(deg: Double): Angle = new Angle(deg * Math.PI / 180.0)

  /**
   * Create angle from gradians.
   */
  def fromGradians(grad: Double): Angle = new Angle(grad * Math.PI / 200.0)

  /**
   * Create angle from turns.
   */
  def fromTurns(turns: Double): Angle = new Angle(turns * 2 * Math.PI)

/**
 * Safe angle operations.
 */
object SafeAngle:

  /**
   * Create angle from radians.
   */
  def radians(rad: Double): Angle = Angle.fromRadians(rad)

  /**
   * Create angle from degrees.
   */
  def degrees(deg: Double): Angle = Angle.fromDegrees(deg)

  /**
   * Create angle from gradians.
   */
  def gradians(grad: Double): Angle = Angle.fromGradians(grad)

  /**
   * Create angle from turns.
   */
  def turns(t: Double): Angle = Angle.fromTurns(t)

  /**
   * Convert degrees to radians.
   */
  def degreesToRadians(deg: Double): Double = deg * Math.PI / 180.0

  /**
   * Convert radians to degrees.
   */
  def radiansToDegrees(rad: Double): Double = rad * 180.0 / Math.PI

  /**
   * Safe arc sine (returns angle).
   */
  def asin(x: Double): Option[Angle] =
    if x < -1.0 || x > 1.0 then None
    else Some(Angle.fromRadians(Math.asin(x)))

  /**
   * Safe arc cosine (returns angle).
   */
  def acos(x: Double): Option[Angle] =
    if x < -1.0 || x > 1.0 then None
    else Some(Angle.fromRadians(Math.acos(x)))

  /**
   * Arc tangent (always valid).
   */
  def atan(x: Double): Angle = Angle.fromRadians(Math.atan(x))

  /**
   * Two-argument arc tangent.
   */
  def atan2(y: Double, x: Double): Angle = Angle.fromRadians(Math.atan2(y, x))

  /**
   * Calculate angle between two 2D vectors.
   */
  def angleBetween(x1: Double, y1: Double, x2: Double, y2: Double): Option[Angle] =
    val dot = x1 * x2 + y1 * y2
    val mag1 = Math.sqrt(x1 * x1 + y1 * y1)
    val mag2 = Math.sqrt(x2 * x2 + y2 * y2)
    val denom = mag1 * mag2
    if denom < 1e-10 then None
    else
      val cosAngle = (dot / denom).max(-1.0).min(1.0)
      Some(Angle.fromRadians(Math.acos(cosAngle)))

  /**
   * Calculate bearing from one point to another (in degrees, 0 = North).
   */
  def bearing(fromX: Double, fromY: Double, toX: Double, toY: Double): Angle =
    val dx = toX - fromX
    val dy = toY - fromY
    val angle = Math.atan2(dx, dy) // Note: atan2(x, y) for bearing
    Angle.fromRadians(angle).normalize

  /**
   * Interpolate between two angles (shortest path).
   */
  def interpolate(a: Angle, b: Angle, t: Double): Angle =
    val factor = t.min(1.0).max(0.0)
    val diff = (b - a).normalizeSigned.radians
    Angle.fromRadians(a.radians + diff * factor)

  /**
   * Find the smallest angle difference.
   */
  def shortestDifference(a: Angle, b: Angle): Angle =
    (b - a).normalizeSigned

  /**
   * Check if an angle is within a range (inclusive).
   */
  def isInRange(angle: Angle, start: Angle, end: Angle): Boolean =
    val normAngle = angle.normalize.radians
    val normStart = start.normalize.radians
    val normEnd = end.normalize.radians

    if normStart <= normEnd then
      normAngle >= normStart && normAngle <= normEnd
    else
      // Range wraps around (e.g., 350 to 10 degrees)
      normAngle >= normStart || normAngle <= normEnd

File: .UNSAFE-ALL-BINDINGS-DELETED-FINAL-20260125/bindings/php/src/Proven/SafeAngle.php
<?php
// SPDX-License-Identifier: PMPL-1.0
// SPDX-FileCopyrightText: 2025 Hyperpolymath

declare(strict_types=1);

namespace Proven;

/**
 * Angle in degrees with safe operations.
 */
readonly class Degrees implements \Stringable
{
    public function __construct(
        public float $value,
    ) {}

    /**
     * Convert to radians.
     */
    public function toRadians(): Radians
    {
        return new Radians($this->value * M_PI / 180.0);
    }

    /**
     * Normalize to [0, 360).
     */
    public function normalize(): self
    {
        $normalized = fmod($this->value, 360.0);
        if ($normalized < 0) {
            $normalized += 360.0;
        }
        return new self($normalized);
    }

    /**
     * Normalize to [-180, 180).
     */
    public function normalizeSigned(): self
    {
        $normalized = fmod($this->value + 180.0, 360.0);
        if ($normalized < 0) {
            $normalized += 360.0;
        }
        return new self($normalized - 180.0);
    }

    /**
     * Add another angle.
     */
    public function add(self $other): self
    {
        return new self($this->value + $other->value);
    }

    /**
     * Subtract another angle.
     */
    public function subtract(self $other): self
    {
        return new self($this->value - $other->value);
    }

    /**
     * Get the shortest difference to another angle.
     */
    public function shortestDifferenceTo(self $other): self
    {
        $diff = fmod($other->value - $this->value + 180.0, 360.0);
        if ($diff < 0) {
            $diff += 360.0;
        }
        return new self($diff - 180.0);
    }

    /**
     * Get sine.
     */
    public function sin(): float
    {
        return sin(deg2rad($this->value));
    }

    /**
     * Get cosine.
     */
    public function cos(): float
    {
        return cos(deg2rad($this->value));
    }

    /**
     * Get tangent.
     *
     * @return Option<float>
     */
    public function tan(): Option
    {
        $cos = $this->cos();
        if (abs($cos) < 1e-10) {
            return Option::none();
        }
        return Option::some(tan(deg2rad($this->value)));
    }

    public function __toString(): string
    {
        return sprintf('%.2f deg', $this->value);
    }
}

/**
 * Angle in radians with safe operations.
 */
readonly class Radians implements \Stringable
{
    public function __construct(
        public float $value,
    ) {}

    /**
     * Convert to degrees.
     */
    public function toDegrees(): Degrees
    {
        return new Degrees($this->value * 180.0 / M_PI);
    }

    /**
     * Normalize to [0, 2*PI).
     */
    public function normalize(): self
    {
        $normalized = fmod($this->value, 2 * M_PI);
        if ($normalized < 0) {
            $normalized += 2 * M_PI;
        }
        return new self($normalized);
    }

    /**
     * Normalize to [-PI, PI).
     */
    public function normalizeSigned(): self
    {
        $normalized = fmod($this->value + M_PI, 2 * M_PI);
        if ($normalized < 0) {
            $normalized += 2 * M_PI;
        }
        return new self($normalized - M_PI);
    }

    /**
     * Add another angle.
     */
    public function add(self $other): self
    {
        return new self($this->value + $other->value);
    }

    /**
     * Subtract another angle.
     */
    public function subtract(self $other): self
    {
        return new self($this->value - $other->value);
    }

    /**
     * Get sine.
     */
    public function sin(): float
    {
        return sin($this->value);
    }

    /**
     * Get cosine.
     */
    public function cos(): float
    {
        return cos($this->value);
    }

    /**
     * Get tangent.
     *
     * @return Option<float>
     */
    public function tan(): Option
    {
        $cos = $this->cos();
        if (abs($cos) < 1e-10) {
            return Option::none();
        }
        return Option::some(tan($this->value));
    }

    public function __toString(): string
    {
        return sprintf('%.4f rad', $this->value);
    }
}

/**
 * Safe angle operations.
 */
class SafeAngle
{
    /**
     * Convert degrees to radians.
     */
    public static function degToRad(float $degrees): float
    {
        return $degrees * M_PI / 180.0;
    }

    /**
     * Convert radians to degrees.
     */
    public static function radToDeg(float $radians): float
    {
        return $radians * 180.0 / M_PI;
    }

    /**
     * Normalize degrees to [0, 360).
     */
    public static function normalizeDegrees(float $degrees): float
    {
        $normalized = fmod($degrees, 360.0);
        if ($normalized < 0) {
            $normalized += 360.0;
        }
        return $normalized;
    }

    /**
     * Normalize radians to [0, 2*PI).
     */
    public static function normalizeRadians(float $radians): float
    {
        $normalized = fmod($radians, 2 * M_PI);
        if ($normalized < 0) {
            $normalized += 2 * M_PI;
        }
        return $normalized;
    }

    /**
     * Linear interpolation between two angles (in degrees).
     * Takes the shortest path around the circle.
     *
     * @param float $t Interpolation factor (0-1)
     */
    public static function lerpAngleDegrees(float $from, float $to, float $t): float
    {
        $t = max(0, min(1, $t));
        $diff = fmod($to - $from + 540.0, 360.0) - 180.0;
        return self::normalizeDegrees($from + $diff * $t);
    }

    /**
     * Linear interpolation between two angles (in radians).
     * Takes the shortest path around the circle.
     *
     * @param float $t Interpolation factor (0-1)
     */
    public static function lerpAngleRadians(float $from, float $to, float $t): float
    {
        $t = max(0, min(1, $t));
        $diff = fmod($to - $from + 3 * M_PI, 2 * M_PI) - M_PI;
        return self::normalizeRadians($from + $diff * $t);
    }

    /**
     * Calculate the shortest angular distance in degrees.
     */
    public static function angularDistanceDegrees(float $from, float $to): float
    {
        $diff = fmod($to - $from + 180.0, 360.0) - 180.0;
        return abs($diff);
    }

    /**
     * Calculate the shortest angular distance in radians.
     */
    public static function angularDistanceRadians(float $from, float $to): float
    {
        $diff = fmod($to - $from + M_PI, 2 * M_PI) - M_PI;
        return abs($diff);
    }

    /**
     * Safe arcsine that returns None for out-of-range inputs.
     *
     * @return Option<float> Radians
     */
    public static function asin(float $x): Option
    {
        if ($x < -1.0 || $x > 1.0) {
            return Option::none();
        }
        return Option::some(asin($x));
    }

    /**
     * Safe arccosine that returns None for out-of-range inputs.
     *
     * @return Option<float> Radians
     */
    public static function acos(float $x): Option
    {
        if ($x < -1.0 || $x > 1.0) {
            return Option::none();
        }
        return Option::some(acos($x));
    }

    /**
     * Safe arctangent (atan2 variant).
     */
    public static function atan2(float $y, float $x): float
    {
        return atan2($y, $x);
    }

    /**
     * Check if an angle (in degrees) is within a range.
     * Handles wrap-around correctly.
     */
    public static function isWithinDegrees(float $angle, float $start, float $end): bool
    {
        $angle = self::normalizeDegrees($angle);
        $start = self::normalizeDegrees($start);
        $end = self::normalizeDegrees($end);

        if ($start <= $end) {
            return $angle >= $start && $angle <= $end;
        }
        // Wrap-around case
        return $angle >= $start || $angle <= $end;
    }

    /**
     * Create degrees from a value.
     */
    public static function degrees(float $value): Degrees
    {
        return new Degrees($value);
    }

    /**
     * Create radians from a value.
     */
    public static function radians(float $value): Radians
    {
        return new Radians($value);
    }

    /**
     * Common angles in degrees.
     */
    public static function rightAngle(): Degrees { return new Degrees(90.0); }
    public static function straightAngle(): Degrees { return new Degrees(180.0); }
    public static function fullRotation(): Degrees { return new Degrees(360.0); }
}
